lib/document.js

```diff
@@ -1149,8 +1149,10 @@ Document.prototype.$set = function $set(path, val, type, options) {
       }
 
       if (utils.isNonBuiltinObject(valForKey) && pathtype === 'nested') {
-        $applyDefaultsToNested(path[key], prefix + key, this);
-        this.$set(prefix + key, path[key], constructing, Object.assign({}, options, { _skipMarkModified: true }));
+        // Create a copy of the object to avoid editing the original
+        const pathKey = utils.clone(path[key]);
```

Comment:
I don't like this change, it seems unnecessarily heavy from a performance perspective. I think we'd just need to refactor and be smarter about setting defaults on the nested path without setting them on the input parameter. We'll take a look and see if we can come up with an alternative solution. I agree this is a bug, I just don't think the solution is ideal.